### ISSUE TITLE
Feature request: support for 2.2" Serial SPI TFT LCD Module Display 240x320 ILI9341

### ISSUE THREAD

**Issue by orangeudav** [MarlinFirmware/Marlin#1102]
Hallo! I have 2.2" Serial SPI TFT LCD Module Display 240x320 ILI9341 IC PCB Adpater SD Socket (http://www.ebay.com/itm/2-2-Serial-SPI-TFT-LCD-Module-Display-240x320-ILI9341-IC-PCB-Adpater-SD-Socket-/171342809726). Any chance to support it (for Rumba in my case)?

**Comment by jgauchia** [comment]
Hi, 
I've found this :

https://code.google.com/p/ucglib/

Support this kind of LCD Modules and the developer is the same than u8glib. Maybe it's possible.

**Comment by boelle** [comment]
how many would benefit from this ?

**Comment by thinkyhead** [comment]
I think it's always best if someone who has the actual device can figure it out and submit the feature as a pull request. Requests like this will tend to hang around in the air without the magic combination of hacking ability, time, and possession of the device. This is the kind of request for which it might help to offer _enticements_ and set a bounty. If someone really wants support badly enough, they can offer to supply the needed equipment, at least.

**Comment by boelle** [comment]
I tend to agree, that is also the reason why i closed a few of the feature requests... either they where too old or too few would benefit from it. will do the same with this one

**Comment by elf128** [comment]
I have the same screen as topic starter and my electronics has been built with this screen in mind initially. I've ported Marlin to support my electronics a long time ago, but there was no extra flash space for LCD functionality. Now I've changed MCU and I have some time to add support for this screen.
Do you have any specific suggestions or demands for LCD code which can make future integration easier.

**Comment by thinkyhead** [comment]
@elf128 Basically, take the existing LCD code as a guide. Most LCD controllers can be supported without too many changes, as they all tend to use the same routines and libraries. The most recent example is the `ELB_FULL_GRAPHICS_CONTROLLER` which didn't require a whole lot of code.

**Comment by AnHardt** [comment]
With 240x320 pixel instead of 128x64 be prepared to make new fonts, a new screen layout, new graphics, ...
Not a task for an afternoon. 
To make your design recognizable as Marlin i'd like to suggest to stay with the extruder, bed, fan- graphics at the top, the statusline at the bottom, and maybe the white stripe in the middle.
Don't try to copy 'dogm_lcd_implementation.h' - it's widely crap.
Try to define "Display_Places" - at least 5 in the top-row. Later populate the places with extruders,fan, chamber, bed in a configuration.

You have about 10 times more pixels. I see memory and timing problems.

**Comment by elf128** [comment]
Yep, that doesn't looks like a picnic. :)
I was thinking of a wrapper for adafruit libraries. Cause a lot of their displays are supported already and they're abstract. So by adding wrapper to generalize an arbitrary adafruit LCD we can open wide range of graphics LCD just by including right library. On the other hand, I've got an experience of porting adafruit library for another LCD an have to say that on the way of generalization they've lost alot of performance. 
On top of all that - It's color display. Frankly, I'm more worried about bandwidth. With this much of a pixels - Asynchronous SPI transfer is a must. And this is an issue already, cause LCD support was written in very arduinish way. :(
We'll see. It wil be fun. :)

P.S. And yes, dogm_\* looks scary.

**Comment by elf128** [comment]
Ok, sorry for super long silence. I was super busy at work.
Screen is working. Sure, it does require a lot of cleaning up before submiting, but at least, now we have something to speak about.
![img_20150823_010520](https://cloud.githubusercontent.com/assets/6124141/9427028/a23e4d00-4933-11e5-8b76-ccd9c5059dde.jpg)

**Comment by Wackerbarth** [comment]
Be sure to move this over to the MarlinDev repository. (It's based on the same root tree) A lot has changed in the interim and features to be included in the future will go in via that path.

**Comment by elf128** [comment]
With multiline layout 25x10 and special symbols
![img_20150823_042058](https://cloud.githubusercontent.com/assets/6124141/9429034/ceb6f38e-4990-11e5-8ebf-b9a72b384bac.jpg)
With high resolution fonts
![img_20150823_042900](https://cloud.githubusercontent.com/assets/6124141/9429040/11cfd820-4991-11e5-97ee-df76d65b4abf.jpg)

**Comment by elf128** [comment]
@Wackerbarth - "Be sure to move this over to the MarlinDev repository. (It's based on the same root tree) A lot has changed in the interim and features to be included in the future will go in via that path."

Probably, I'll require some assistance for that. Cause back in a very begining, when I start my work I modified a lot of stuff to make sure my electronics is working. But when I decide to add this LCD support, I noticed that it's being a year since my last merging with Marlin code deport. Sure enought it was about of week of sequential merging to catch up with you guys.
Now, my biggest question is about naming you're using for diferent branches. When you checking out root of Marlin tree, you have "Branches", "Tags" and "Trunk" and names of the branches inside Tags and Braches is not consistant with names I can see over web.
Do you have any documentation about what is what?

**Comment by thinkyhead** [comment]
@elf128 The current README gives a little overview of the current branches. But basically, just make a fork of MarlinDev, copy MarlinDev/master to a new feature branch, and start your work there. When you make your PR submit it against MarlinDev/master.

**Comment by vivian-ng** [comment]
Sorry for digging up an old thread, but has there been any progress on adding ILI9341 support to Marlin? I can't seem to find any specific file in the code referencing ILI9341.

**Comment by mduca** [comment]
I've had success using this LCD with an ESP8266. Would love for it to work with Marlin. Not exactly sure how to go about doing that.

**Comment by vivian-ng** [comment]
I wonder if anyone has any hints for adding a new display to Marlin. Like, is there a CPP or header file which I can reference to see how I can write a module to support the ILI9341 for use with Marlin?

**Comment by AnHardt** [comment]
Currently there are two kinds of displays supported by Marlin.
**Caracter** displays based on HITACHI HD44780 controllers and
**graphical** **monochrome** **128x64** displays drivable by u8g-lib.
ILI9341 is a **RGB** **240 x 320** display. Supporting it would need the integration of ucg-lib or similar plus creating graphics, fonts and screens suitable for the much higher resolution with a different aspect ratio.
That's perfectly doable but needs a mayor efford and a deep insight of what is currently done in Marlin. (Not a Sunday afternoon project - more some month of work)

Sorry for repeating myself.

**Comment by vivian-ng** [comment]
@AnHardt thank you for the brief explanation. It seems to be a major undertaking, hopefully my interest does not die out halfway.
There is a lcdprint_u8g.cpp in the bugsfix-2.0.x branch which seems to be a wrapper for display functions around the u8glib library. There is also a malyanlcd.cpp which added support for the Malyan LCD. There are also a bunch of other files in the lcd subdirectory. It seems that getting something onto the screen is handled by lcdprint.h (and a corresponding CPP file like lcdprint_u8g.cpp for a particular device driver), but the actual support for a particular device is through an implementation of ultralcd.h. Are these the files which I should be working through if I really want to add support for a new display?

**Comment by Eyolon** [comment]
hi 
i'm interested to continue the work around the ili9341. @elf128 : i've saw your pictures and i want to begin with your last commit with this result or more if you have.

**Comment by ejtagle** [comment]
Well, i am pretty close to get the ili9331 / ili9341 / ili9481 TFT screens working with Marlin.

Already done:
-A new font importer with associated compressor, supporting alpha blending: For example, a 14x17 proportional font (OpenSans Italic), that would take uncompressed 9011 bytes, after compression, takes 3545 bytes. The compression is loosely based on PNG, but with no SRAM usage (basically, the PNG predictors + rle + adaptive golomb bit coding)
-A new bitmap importer for truecolor images with alpha channel, also using PNG like compression (PNG predictors/rle/adaptive golomb bit coding) - Compression performance similar or better than PNG.
-All the associated assembly optimized decompressors for those formats for AVR (and also working on ARM and any 32bit board, where the default C version is more than enough)
-A new TFT library that allows to READ the TFT memory, superoptimized for speed (reading is required to implement alpha blending of fonts and images)
-All the usual graphics primitives, as supported by u8glib, but targeting color display.
-Most primitives can be filled with horizontal and vertical color gradients.
-Rectangular clipping for all primitives
-All that nicely packed into a TFT class, mostly compatible with adafruit libraries

Then, a new "windowing" system (not truly a window manager, as that would require lots of SRAM)
But the GUI objects allow to describe panels of controls with dynamic values. Already done are Buttons, Checkboxes, Images, Labels, Lists, ProgressBars, RadioButtons, Scrollbars, Sliders, Radio buttons, and Toggle controls.

All this takes 30 bytes of SRAM (+stack). All the configuration and layout resides in FLASH. 

https://youtu.be/yhVwXYrp9qg
That small demo takes 78kbytes of FLASH, and it shows the time that takes to paint the entire UI from a reset (when the screen turns white). And also shows how responsive it is. The demo is running on an atmega2560. 

On any 32bit Arduino, the screen creation and update is completely instantaneous

**Comment by elf128** [comment]
Oh, sorry. I just saw new coments on this thread.
Yes. I've menaged to make it work with marlin and this code is working for like 4 years already and not giving me any troubles ever since.
I've never submitted my code because of following reasons.

People complettely missjudging this issue. The problem is not a screen.

I mean, getting screen running is a trivial task. It's just sequence of fixed data over SPI. Then, it's a stream of pixels. Yes you have to deal with fonts and other small things, but it's not complicated, it's just tedious.

Even writing everyhing from scratch is a picnic in a park in comparison with integration of that into Marlin.

For a starter, you need a buffer for text, cause Marlin expect screen to be character based. It simply tryes to refresh whole screen every tick, and if it takes too long, main loop goes nuts. You need to update screen selectively. And that's where your RAM budget goes through the roof, cause you need two screen buffers to make it work fast on 16MHz device.
Also, there are tonnes of hardcoded stuff in ALL screen related parts of the code. UI is pretty much one single hardcoded piece, cause it was writtent for one screen, and then totally obfuscated during adding support of the second screen. 
Long story short, By the time I've finished integration, I've touced almost half of the code base and added couple of adstraction layers and rewrote UI to work through it. 
However, by the time I've finished, so many things have been submitted and I just gave up merging those. Also, taking into account overal amount of changes, I'm almost certanly broke compatibility with other screens, it should work, but I simply have no way to test the code. 

I'm pretty sure this is what happen with all attempts to add support of any other screen. People simply end up in their own island of code which is too far away from original Marlin code base to merge back.

**Comment by AnHardt** [comment]
3 times locked but still commentable?

It's done.

**Comment by thisiskeithb** [comment]
> 3 times locked but still commentable?
> 
> It's done.

It's a github bug. The maintainers can't even lock this thread (and other older ones), but github has been made aware of the issue.

**Comment by AnHardt** [comment]
Could it work if the FR-label is removed?

**Comment by shitcreek** [comment]
This is a gitHub bug I believe. I've already gotten in touch with support, and they've escalated the issue to their engineers.
